src/dune_engine/build_system.ml

```diff
@@ -179,6 +189,18 @@ module State = struct
   let reset_errors () = Svar.write errors Error.Set.empty
 
   let add_error error =
+    let current = Svar.read t in
+    let open Fiber.O in
+    let* () =
+      match current with
+      | Building current ->
+        Svar.write t
+          (Building
+             { current with
+               number_of_rules_failed = current.number_of_rules_failed + 1
+             })
+      | _ -> assert false
```

Comment:
Shall we factor out this repeated pattern? I supposed this was copied from `start_rule_exn` above.

Reply:
It is a different count being incremented so there is nothing to factor. This was sitting in its own function in Svar before but I later inlined it since it is only used here.

Reply:
Everything apart from the actual record update can be factored out, right? All this `Svar.read` and `assert false` business doesn't need to be repeated twice.

Reply:
Point taken, thanks!